File: py/src/practise150/p16_longest_substring_without_duplicates.py
'''
Given a string s, find the length of the longest substring without duplicate characters.
'''

class Solution:
    def longestsubstring(self, s):
        length, left, right = 0,0,0
        
        setstring = set()
        while right < len(s):
            while s[right] in setstring:
                setstring.remove(s[left])
                left +=1
            setstring.add(s[right])
            length = max(length, right - left + 1)
            right += 1
        return length

File: go/practise150/p16_longest_substring_without_duplicates.go
package practise150

// Given a string s, find the length of the longest substring without duplicate characters.

func LongestSubstring(s string) int {
	// Dont use this. Use the one with two pointers.
	hashmap := make(map[rune]struct{})
	
	maxlength := 0
	length := 0
	for _, rn := range s {
		if _, ok := hashmap[rn]; !ok {
			hashmap[rn] = struct{}{}
			length += 1
		} else {
			clear(hashmap)
			hashmap[rn] = struct{}{}
			length = 1
		}
		maxlength = max(length, maxlength)
	}
	return maxlength
}

func LongestSubstringWithTwoPointers(s string) int{
	i, j, length := 0, 0, 0
	
	hashset:= make(map[rune]struct{})
	
	for j < len(s) {
		if _, ok := hashset[rune(s[j])]; !ok {
			hashset[rune(s[j])] = struct{}{}
			length = max(length, j - i + 1)
			j += 1
		} else {
			delete(hashset, rune(s[i]))
			i += 1
		}
	}
	return length
}